### PR TITLE
Issue 928 filter in ingress

### LINKED ISSUE #928: Prefilter in broker ingress
**Problem**
Do not persist messages that don't match any triggers.
Also consider tagging events with filter result, which can be used by fanout and retry.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?

**Exit Criteria**
A measurable (binary) test that would indicate that the problem has been resolved.

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by tayarani**
Fixes #928 

## Proposed Changes
Pre-filter events at the ingress level when there are no targets interested in it. This will help save pub/sub resources by testing each event against all target tilters *before* sending the event to pub/sub.

### CODE REVIEW

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:73]: I have thought about the positioning of this block wrt to the block under it.

Right now if a broker is setup (even incorrectly, e.g., no topic) without targets, ingress will accept and acknowledge all incoming events.

If we move this block under the one below, it will at least check to see the topic exists before accepting events.

I'm open to either approach, and would appreciate it if anyone else has other thoughts.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:107]: IMO it will be more readable to have `hasTrigger := false` instead.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:60]: The `Targets` and `DecoupleQueue` fields can be in their own new line.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:107]: In go all variables are zero initialized: https://golang.org/ref/spec#The_zero_value.

Comment (zhongduo) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:107]: Yes, I do know that, that's why I said readable not bug :). IMO [short variable declaration](https://golang.org/ref/spec#Short_variable_declarations) is a cleaner and easier to understand syntax, especially for these primary types.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:111]: is this required? It seems to me that without this the function will still return true later.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:320]: `hasTrigger` can be inlined into the `if` statement.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:360]: Not sure if this complexity is needed here, since you always create a new server, is there any chance that there will be topic already in the test server? If so you should handle `Exists` properly. Otherwise you can safely call CreateTopic.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:102]: If this is for testing purpose only, do you have to export the name?

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:320]: Any reason you use `context.Background()` instead of `ctx`?

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:317]: You should never used `fmt.Printf` in test code except for your own debugging purpose. Use `t.Logf` instead.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:111]: That was intended to be false. It's for early breaking out of the 'RangeAllTargets' loop. Fixed.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:320]: Good catch.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:360]: I took this block of code from 'TestMultiTopicDecoupleSink'. Seems like the correct way to ensure the test is setup properly.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:317]: Left over from manual debugging. Removed.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (liu-cong) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:73]: I think putting this below the `topic, err := m.getTopicForBroker(ctx, broker)` is better. `getTopicForBroker` should usually be a faster operation than applying the filters.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (liu-cong) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:108]: I think we should gate this feature behind a flag and disable it by default. Alternatively we can make sure we do some performance testing before we release. However, the latter could potentially delay our release.

Thoughts? @grantr

Comment (grantr) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:108]: IMO we don't need a flag for this, and we should test the performance impact anyway.

If we're concerned about filtering taking too long in edge cases, we can add a timeout after which the event is sent to the decoupling topic (and a prefilter timeout metric is incremented). This would be similar to fanout's initial delivery timeout to retry (but with a much shorter timeout, maybe 100ms).

Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:124]: As discussed, I added a feature flag for this until we test the performance change. I've added TODOs to clean up the feature gate as part of #1804.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go]: Done.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (zhongduo) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:108]: It only works when the user is in the same package.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:240]: You might want to add a similar test but with an empty map (not nil).

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:300]: `defer psClient.close()`

Comment (zhongduo) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:300]: Alternatively, you can use https://github.com/google/knative-gcp/blob/eb5ac1b2f0231ba7bb1681858f3478a001938341/pkg/reconciler/testing/pstest.go#L200

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (zhongduo) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:60]: I don't think line length is the concern here, in our code base, you do normally have multiple line for struct initialization. That's why `Brokers:` is in its own line.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:240]: Done

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (grantr) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:278]: Can you add a test of multiple targets with none matching?

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (grantr) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:330]: Add a `TODO(#1804)` here to delete lines 330-338

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (grantr) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:339]: Can we remove the mock and check whether the message made it to the topic or not? Then we test that `EventFilterFunc` is set correctly in the tested code. (Arguably E2E tests would also cover this, but having it in a unit test makes identifying a bug much quicker)

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (grantr) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:396]: I'm fine with using a mock here for the purpose of the test, but if we're changing the above test it might be useful to make this one consistent. Consider this suggestion optional since I expect this is a temporary test.

File: pkg/broker/handler/processors/filter/processor.go
Comment (grantr) [inline pkg/broker/handler/processors/filter/processor.go:98]: Making this an exported function seems a bit odd to me in terms of encapsulation and modularity. Don't take this as an objection though. We may want to consider a followup that refactors the ingress to use processors in the same style as fanout and retry.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (grantr) [inline pkg/broker/ingress/multi_topic_decouple_sink_test.go:43]: Can you elaborate on this comment to explain why adding a target ensures events are sent to pubsub? The reader here may not be familiar with the ingress filtering behavior.

**Review by grantr — COMMENTED** [review]
Thanks @tayarani! I have a few nits on tests and docs and other non-blocking comments.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (yolocs) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:339]: This change would potentially make e2e test harder for filtering in fanout/retry. Today, we just need to send an event that doesn't match any trigger and expect not receiving it. We know for sure it's the filtering in fanout/retry makes it (not) happen. But with this change, the event would be filtered on ingress.

**Comment by yolocs** [comment]
We don't have to wait until this gets merged to know the performance impact. I think a micro benchmark should do it. @ian-mi has done this for multiple critical parts, e.g. https://github.com/google/knative-gcp/blob/master/pkg/broker/handler/processors/deliver/processor_test.go#L367

I don't mean to block this change. But I would prefer getting a rough grasp on the performance impact upfront (instead of merging it and later find out it has a huge impact and revert the change).

**Comment by grantr** [comment]
Good idea to add a micro benchmark @yolocs. @tayarani does that seem reasonable to add to this PR?

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (Harwayne) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:92]: I don't think this needs a metric. 

It might be helpful for us to be able to say 'look how much we saved in money and latency (by sending the upstream ack faster)', but I don't see much utility for users, other than when investigating different Broker implementations.

If users are concerned that events are being dropped when they shouldn't be, then I think our answer is that you need to compare Broker `event_count` and Trigger `event_count`. Because the general case of the problem is 'I am sending events to the Broker, but my Trigger is not receiving them'. And that is a problem regardless of whether other Triggers happen to include that event or not. So distinguishing between it being dropped here or at Fanout shouldn't matter.

If you think it is likely this area will have a bug, then a metric may make it easier to track down. But I would argue adding a debug level log statement would likely be better, as we could include the attributes of the event being dropped and investigate if we think it was a mistake or not.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (Harwayne) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:117]: How will this be set?

IIUC, this Deployment is created by the BrokerCell reconciler. How will that reconciler choose to set or not set this env var?

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (Harwayne) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:123]: I agree with @zhongduo, explicitly setting to false is quicker to understand.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (Harwayne) [inline pkg/broker/ingress/multi_topic_decouple_sink.go:107]: Can be injected instead?

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:92]: Thanks for weighing in. I added a log message to help debugging just in case it's needed.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:108]: Can u elaborate please?

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go:280]: Sure. done.

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go]: Good point. I have added annotations to the broker cell now which will set this env variable on ingress pods if they are set.

**Comment by tayarani** [comment]
Micro benchmark results:

Running 3 times with the feature *disabled* (i.e., current default behaviour):
```pkg: github.com/google/knative-gcp/pkg/broker/ingress
BenchmarkIngressHandler/0_bytes-12         	  119655	      9726 ns/op
BenchmarkIngressHandler/1000_bytes-12      	   95484	     17588 ns/op
BenchmarkIngressHandler/100000_bytes-12    	    8024	    130719 ns/op
PASS
ok  	github.com/google/knative-gcp/pkg/broker/ingress	12.013s

pkg: github.com/google/knative-gcp/pkg/broker/ingress
BenchmarkIngressHandler/0_bytes-12         	  103238	      9874 ns/op
BenchmarkIngressHandler/1000_bytes-12      	  108384	     16846 ns/op
BenchmarkIngressHandler/100000_bytes-12    	    7946	    126279 ns/op
PASS
ok  	github.com/google/knative-gcp/pkg/broker/ingress	12.782s

pkg: github.com/google/knative-gcp/pkg/broker/ingress
BenchmarkIngressHandler/0_bytes-12         	  120055	     10090 ns/op
BenchmarkIngressHandler/1000_bytes-12      	   96837	     17096 ns/op
BenchmarkIngressHandler/100000_bytes-12    	    8575	    365640 ns/op
PASS
ok  	github.com/google/knative-gcp/pkg/broker/ingress	12.959s```

**Comment by tayarani** [comment]
New benchmark results with updated benchmark test which sweeps target counts as well:

With the feature disabled (i.e., current default behaviour):
```
pkg: github.com/google/knative-gcp/pkg/broker/ingress
BenchmarkIngressHandler/event_size_0_bytes_-_1_targets-12         	  116500	     10273 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_1_targets-12      	  106446	     10558 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_1_targets-12    	    7412	    363537 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_5_targets-12         	  114288	     10198 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_5_targets-12      	   95527	     12822 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_5_targets-12    	   10372	    108859 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_10_targets-12        	  120920	     10303 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_10_targets-12     	  120037	     17044 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_10_targets-12   	   11582	    107944 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_50_targets-12        	  114660	     10488 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_50_targets-12     	  111990	     12593 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_50_targets-12   	   11056	    101499 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_100_targets-12       	  108790	     10072 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_100_targets-12    	  112933	     17784 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_100_targets-12  	    8248	    621597 ns/op
PASS
ok  	github.com/google/knative-gcp/pkg/broker/ingress	37.846s
```

With the feature enabled (i.e., new behaviour):
```
pkg: github.com/google/knative-gcp/pkg/broker/ingress
BenchmarkIngressHandler/event_size_0_bytes_-_1_targets-12         	  115342	     10853 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_1_targets-12      	   73153	     15202 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_1_targets-12    	    7351	    415066 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_5_targets-12         	   86582	     12181 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_5_targets-12      	   80860	     18619 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_5_targets-12    	   10368	    115420 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_10_targets-12        	   79036	     13373 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_10_targets-12     	   91483	     17765 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_10_targets-12   	   11096	    299258 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_50_targets-12        	   49312	     23270 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_50_targets-12     	   53337	     21697 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_50_targets-12   	    7082	    409847 ns/op
BenchmarkIngressHandler/event_size_0_bytes_-_100_targets-12       	   33829	     41354 ns/op
BenchmarkIngressHandler/event_size_1000_bytes_-_100_targets-12    	   31771	     34050 ns/op
BenchmarkIngressHandler/event_size_100000_bytes_-_100_targets-12  	    6237	    261891 ns/op
PASS
ok  	github.com/google/knative-gcp/pkg/broker/ingress	37.085s
```

File: pkg/broker/handler/processors/filter/processor.go
Comment (tayarani) [reply pkg/broker/handler/processors/filter/processor.go:98]: I do believe the code that decides if an event should be sent out by fan out, should be the exact same as the code that decides to drop events at ingress. Any discrepancy between those two methods would be problematic. So I opted to re-use the same method.

We could refactor into a common place where they both call a helper if that helps.

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go]: I decided to keep the mock. If we remove it, testing the case where an event shouldn't make it down stream is difficult with current pub/sub. It would make the tests overly complicated imo. so i didn't spend any more time on it. Hope that's okay

File: pkg/broker/ingress/multi_topic_decouple_sink_test.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink_test.go]: Done

File: pkg/broker/ingress/multi_topic_decouple_sink.go
Comment (tayarani) [reply pkg/broker/ingress/multi_topic_decouple_sink.go:108]: I did consider that, but it seemed a little more complicated than the var. I'm open to it if we feel the function ptr won't work.

**Comment by grantr** [comment]
According to [benchstat](https://godoc.org/golang.org/x/perf/cmd/benchstat) the difference is not statistically significant in any of the benchmarks. I'm not sure exactly how that's calculated though, and in some cases it seems pretty significant to me (event_size_100000_bytes_-_5_targets-12 is 109us vs 326us). I'm fine with calling it equivalent performance, though anyone with more expertise in microbenchmarks please feel free to disagree.

```
$ benchstat nofilter.txt prefilter.txt
name                                                     old time/op  new time/op  delta
IngressHandler/event_size_0_bytes_-_1_targets-12         10.3µs ± 0%  10.3µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_1000_bytes_-_1_targets-12      10.6µs ± 0%  16.0µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_100000_bytes_-_1_targets-12     364µs ± 0%   132µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_0_bytes_-_5_targets-12         10.2µs ± 0%  11.6µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_1000_bytes_-_5_targets-12      12.8µs ± 0%  13.7µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_100000_bytes_-_5_targets-12     109µs ± 0%   326µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_0_bytes_-_10_targets-12        10.3µs ± 0%  12.6µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_1000_bytes_-_10_targets-12     17.0µs ± 0%  12.8µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_100000_bytes_-_10_targets-12    108µs ± 0%   411µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_0_bytes_-_50_targets-12        10.5µs ± 0%  22.2µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_1000_bytes_-_50_targets-12     12.6µs ± 0%  23.3µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_100000_bytes_-_50_targets-12    101µs ± 0%   114µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_0_bytes_-_100_targets-12       10.1µs ± 0%  39.0µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_1000_bytes_-_100_targets-12    17.8µs ± 0%  41.3µs ± 0%   ~     (p=1.000 n=1+1)
IngressHandler/event_size_100000_bytes_-_100_targets-12   622µs ± 0%   698µs ± 0%   ~     (p=1.000 n=1+1)

**Review by grantr — APPROVED** [review]
Nice work @tayarani!

/lgtm
/hold
Holding in case others who reviewed have comments.

**Comment by liu-cong** [comment]
I am going to cancel the hold since this has been approved for a week.

/unhold

**Pull request merged**
